JSON parser panics on certain malformed input

Originally reported over in hashicorp/terraform#23964.

Given the following input, the JSON parser panics:

```json
{
  "http_proxy_url: "http://xxxxxx",
}
```

Note that the property name string is not closed, and so presumably the parser is still in the string literal parsing state when it reaches the end of the buffer.

The panic stack trace reported in the Terraform issue is:

```
panic: runtime error: index out of range [69] with length 69

goroutine 1 [running]:
github.com/hashicorp/hcl/v2/json.(*peeker).Read(...)
        /private/tmp/terraform-20191211-59618-1xdshdv/terraform-0.12.18/pkg/mod/github.com/hashicorp/hcl/v2@v2.0.0/json/peeker.go:20
github.com/hashicorp/hcl/v2/json.parseObject.func1(0x0, 0xc000232468, 0x1, 0x266, 0x7ffeefbff7da, 0x1a, 0x12, 0x1a, 0x468, 0x12, ...)
        /private/tmp/terraform-20191211-59618-1xdshdv/terraform-0.12.18/pkg/mod/github.com/hashicorp/hcl/v2@v2.0.0/json/parser.go:127 +0xd8
github.com/hashicorp/hcl/v2/json.parseObject(0xc0005351e8, 0xc000207b60, 0x7ffeefbff7da, 0x1a, 0x12, 0x3)
        /private/tmp/terraform-20191211-59618-1xdshdv/terraform-0.12.18/pkg/mod/github.com/hashicorp/hcl/v2@v2.0.0/json/parser.go:157 +0x16d9
github.com/hashicorp/hcl/v2/json.parseValue(0xc0005351e8, 0x67, 0x267, 0x7ffeefbff7da, 0x1a, 0x12)
        /private/tmp/terraform-20191211-59618-1xdshdv/terraform-0.12.18/pkg/mod/github.com/hashicorp/hcl/v2@v2.0.0/json/parser.go:45 +0x708
github.com/hashicorp/hcl/v2/json.parseFileContent(0xc000232000, 0x4ce, 0x6ce, 0x7ffeefbff7da, 0x1a, 0x0, 0xc000535400, 0x203000, 0xc000535398, 0x10d4636)
        /private/tmp/terraform-20191211-59618-1xdshdv/terraform-0.12.18/pkg/mod/github.com/hashicorp/hcl/v2@v2.0.0/json/parser.go:21 +0x133
github.com/hashicorp/hcl/v2/json.Parse(0xc000232000, 0x4ce, 0x6ce, 0x7ffeefbff7da, 0x1a, 0x0, 0x0, 0xc0004ec9b2, 0x9)
        /private/tmp/terraform-20191211-59618-1xdshdv/terraform-0.12.18/pkg/mod/github.com/hashicorp/hcl/v2@v2.0.0/json/public.go:21 +0x77
github.com/hashicorp/terraform/command.(*Meta).addVarsFromFile(0xc0000f0700, 0x7ffeefbff7da, 0x1a, 0x4e, 0xc00026b770, 0x0, 0x0, 0x0)
        /private/tmp/terraform-20191211-59618-1xdshdv/terraform-0.12.18/src/github.com/hashicorp/terraform/command/meta_vars.go:158 +0x59c
github.com/hashicorp/terraform/command.(*Meta).collectVariableValues(0xc0000f0700, 0xc0000aa730, 0x0, 0x0, 0xc000229930)
        /private/tmp/terraform-20191211-59618-1xdshdv/terraform-0.12.18/src/github.com/hashicorp/terraform/command/meta_vars.go:112 +0x3a9
github.com/hashicorp/terraform/command.(*ApplyCommand).Run(0xc0000f0700, 0xc00003a090, 0x5, 0x5, 0xc00006e000)
        /private/tmp/terraform-20191211-59618-1xdshdv/terraform-0.12.18/src/github.com/hashicorp/terraform/command/apply.go:195 +0xd23
github.com/mitchellh/cli.(*CLI).Run(0xc0000ce640, 0xc0000ce640, 0xc000535cc0, 0x1)
        /private/tmp/terraform-20191211-59618-1xdshdv/terraform-0.12.18/pkg/mod/github.com/mitchellh/cli@v1.0.0/cli.go:255 +0x1da
main.wrappedMain(0x0)
        /private/tmp/terraform-20191211-59618-1xdshdv/terraform-0.12.18/src/github.com/hashicorp/terraform/main.go:238 +0xc44
main.realMain(0x0)
        /private/tmp/terraform-20191211-59618-1xdshdv/terraform-0.12.18/src/github.com/hashicorp/terraform/main.go:102 +0xb4
main.main()
        /private/tmp/terraform-20191211-59618-1xdshdv/terraform-0.12.18/src/github.com/hashicorp/terraform/main.go:38 +0x3a
```

In the situation where this was reported, Terraform was parsing a `.tfvars.json` file.

--- a/json/public_test.go
+++ b/json/public_test.go
@@ -1,6 +1,7 @@
 package json
 
 import (
+	"strings"
 	"testing"
 
 	"github.com/hashicorp/hcl/v2"
@@ -93,5 +94,21 @@ func TestParseTemplateUnwrap(t *testing.T) {
 
 	if !val.RawEquals(cty.True) {
 		t.Errorf("wrong result %#v; want %#v", val, cty.True)
+	}
+}
+
+func TestParse_malformed(t *testing.T) {
+	src := `{
+  "http_proxy_url: "http://xxxxxx",
+}`
+	file, diags := Parse([]byte(src), "")
+	if got, want := len(diags), 2; got != want {
+		t.Errorf("got %d diagnostics; want %d", got, want)
+	}
+	if err, want := diags.Error(), `Missing property value colon`; !strings.Contains(err, want) {
+		t.Errorf("diags are %q, but should contain %q", err, want)
+	}
+	if file == nil {
+		t.Errorf("got nil File; want actual file")
 	}
 }

--- a/json/scanner.go
+++ b/json/scanner.go
@@ -107,6 +107,15 @@ func scan(buf []byte, start pos) []token {
 			})
 			// If we've encountered an invalid then we might as well stop
 			// scanning since the parser won't proceed beyond this point.
+			// We insert a synthetic EOF marker here to match the expectations
+			// of consumers of this data structure.
+			p.Pos.Column++
+			p.Pos.Byte++
+			tokens = append(tokens, token{
+				Type:  tokenEOF,
+				Bytes: nil,
+				Range: posRange(p, p),
+			})
 			return tokens
 		}
 	}

--- a/json/scanner_test.go
+++ b/json/scanner_test.go
@@ -829,6 +829,21 @@ func TestScan(t *testing.T) {
 						},
 					},
 				},
+				{
+					Type: tokenEOF,
+					Range: hcl.Range{
+						Start: hcl.Pos{
+							Byte:   1,
+							Line:   1,
+							Column: 2,
+						},
+						End: hcl.Pos{
+							Byte:   1,
+							Line:   1,
+							Column: 2,
+						},
+					},
+				},
 			},
 		},
 	}

